src/persist-client/src/iter.rs

```diff
@@ -45,48 +48,10 @@ fn clone_tuple<T, D>((k, v, t, d): TupleRef<T, D>) -> Tuple<T, D> {
     ((k.to_vec(), v.to_vec()), t, d)
 }
 
-/// The data needed to fetch a batch part, bundled up to make it easy
-/// to send between threads.
-#[derive(Debug)]
-pub(crate) struct FetchData<T> {
-    shard_id: ShardId,
-    blob: Arc<dyn Blob + Send + Sync>,
-    metrics: Arc<Metrics>,
-    read_metrics: fn(&BatchPartReadMetrics) -> &ReadMetrics,
-    shard_metrics: Arc<ShardMetrics>,
-    part_key: PartialBatchKey,
-    part_desc: Description<T>,
-}
-
-impl<T: Codec64 + Timestamp + Lattice> FetchData<T> {
-    async fn fetch(self) -> anyhow::Result<EncodedPart<T>> {
-        let Self {
-            shard_id,
-            blob,
-            metrics,
-            read_metrics,
-            shard_metrics,
-            part_key,
-            part_desc,
-        } = self;
-        metrics.compaction.parts_prefetched.inc();
-        fetch_batch_part(
-            &shard_id,
-            &*blob,
-            &metrics,
-            &shard_metrics,
-            read_metrics(&metrics.read),
-            &part_key,
-            &part_desc,
-        )
-        .await
-    }
-}
-
-#[derive(Debug)]
 pub(crate) enum ConsolidationPart<T, D> {
     Queued {
-        data: Option<FetchData<T>>,
+        future: Pin<Box<dyn Future<Output = anyhow::Result<EncodedPart<T>>> + Send>>,
```

Comment:
this is a bit of a bummer, I really liked the FetchData structure before. (haven't tried this locally, so possibly misguided, but) can we get back to something more like that if we split ConsolidationPart::Queued into two leased and non-leased variants? (thinking either replace it with two variants or, if this is easier with the prefetch stuff, a new QueuedPart enum with two variants)

Reply:
> can we get back to something more like that [...]

Almost certainly, yeah!

Though could you say more about why you prefer the old style?

Reply:
the biggest part of my preference is just that I try to avoid `Pin<Box<dyn Future ...>>>` when I can :D. it's a lot easier to follow code around when you can just click through to FetchData instead of having to track down all the places where some dyn trait gets filled in

other random inputs:
- it's a larger change to the code, which means I have to load more context back up to review it
- i'm also hoping it'll let us avoid the `if let Some((c_part @ ConsolidationPart::Queued { .. }, size))` + `let ConsolidationPart::Queued ... else { panic }` construction below, which is a little clunky
- the `dyn Future` means we start going through vcalls when polling. there's essentially no way this will move the needle in context (we're polling calls out to external systems), so it's certainly fine from a performance perspective, but like... when I read a new codebase for the first time, this is the sort of thing that would make me stop and think for a second before moving on. certainly not the biggest factor

are you asking out of curiosity or because you prefer it this way?

Reply:
Thank you!

> are you asking out of curiosity or because you prefer it this way?

Kind of both! Reasons include:
- I have a fairly strong bias to "having less stuff", so I liked that I no longer needed a new concept / intermediate representation.
- It means the only thing you can do with a `ConsolidationPart::Queued` is fetch it, which makes the consolidator easier to reason about.

But also it's fine either way! I just wanted to make sure I understood the motives. I've pushed up a commit that brings back `FetchData` in enum form; let me know what you think!

> i'm also hoping it'll let us avoid [horrible thing]

I _think_ this is orthogonal. This was a replacement for the `Option::take` in my original PR, and you'll notice I have a third mediocre approach to the exact some problem in the new commit. I haven't really loved any of them so far, but I keep trying. 😓 Anyways if you'd prefer a version with both the `Option::take` and the futures, we can do that also!

Reply:
This looks great!